File: packages/react-native/android/src/main/java/com/rejourney/recording/RejourneyNetworkEventFilter.kt
/**
 * Copyright 2026 Rejourney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rejourney.recording

import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import java.net.URI
import java.util.LinkedHashSet

object RejourneyNetworkEventFilter {
    private val internalPathPrefixes = listOf(
        "/api/sdk/config",
        "/api/ingest",
        "/upload/artifacts"
    )
    private const val maxRegisteredInternalUrls = 200

    @Volatile
    private var apiBasePath: String = normalizeBasePath("https://api.rejourney.co")
    private val registeredInternalUrls = LinkedHashSet<String>()

    fun configure(apiUrl: String?) {
        apiBasePath = normalizeBasePath(apiUrl ?: "https://api.rejourney.co")
        synchronized(registeredInternalUrls) {
            registeredInternalUrls.clear()
        }
    }

    fun registerInternalUrl(url: String?) {
        val normalized = normalizeComparableUrl(url) ?: return
        synchronized(registeredInternalUrls) {
            if (registeredInternalUrls.add(normalized)) {
                while (registeredInternalUrls.size > maxRegisteredInternalUrls) {
                    val first = registeredInternalUrls.iterator().next()
                    registeredInternalUrls.remove(first)
                }
            }
        }
    }

    fun shouldIgnore(url: HttpUrl): Boolean {
        if (isRegisteredInternalUrl(url.toString())) {
            return true
        }
        return shouldIgnorePath(url.encodedPath)
    }

    fun shouldIgnore(url: String?): Boolean {
        if (url.isNullOrBlank()) return false
        if (isRegisteredInternalUrl(url)) {
            return true
        }
        return shouldIgnorePath(pathForUrl(url))
    }

    fun shouldIgnore(details: Map<String, Any>): Boolean {
        val url = details["url"] as? String
        if (shouldIgnore(url)) return true

        val path = details["urlPath"] as? String
        return path != null && shouldIgnorePath(path)
    }

    private fun shouldIgnorePath(path: String): Boolean {
        val normalizedPath = normalizePath(path)
        val basePath = apiBasePath
        val prefixes = if (basePath.isBlank()) {
            internalPathPrefixes
        } else {
            internalPathPrefixes + internalPathPrefixes.map { prefix -> normalizePath("$basePath$prefix") }
        }

        return prefixes.any { prefix ->
            normalizedPath == prefix || normalizedPath.startsWith("$prefix/")
        }
    }

    private fun pathForUrl(value: String): String {
        value.toHttpUrlOrNull()?.let { return it.encodedPath }
        return try {
            URI(value).rawPath ?: value.substringBefore('?')
        } catch (_: Exception) {
            value.substringBefore('?')
        }
    }

    private fun isRegisteredInternalUrl(value: String): Boolean {
        val normalized = normalizeComparableUrl(value) ?: return false
        return synchronized(registeredInternalUrls) {
            registeredInternalUrls.contains(normalized)
        }
    }

    private fun normalizeBasePath(value: String): String {
        val trimmed = value.trim()
        if (trimmed.isBlank()) return ""
        val parsed = trimmed.toHttpUrlOrNull() ?: "https://$trimmed".toHttpUrlOrNull()
        val path = parsed?.encodedPath
            ?: try {
                URI(trimmed).rawPath ?: ""
            } catch (_: Exception) {
                ""
            }
        return if (path.isBlank() || path == "/") "" else normalizePath(path)
    }

    private fun normalizePath(value: String): String {
        val withLeadingSlash = if (value.startsWith("/")) value else "/$value"
        val trimmed = withLeadingSlash.trim('/')
        return if (trimmed.isBlank()) "/" else "/$trimmed"
    }

    private fun normalizeComparableUrl(value: String?): String? {
        val trimmed = value?.trim()?.takeIf { it.isNotBlank() } ?: return null
        val parsed = trimmed.toHttpUrlOrNull()
        return parsed?.newBuilder()?.fragment(null)?.build()?.toString()
            ?: trimmed.substringBefore('#')
    }
}

File: packages/core/swift/Recording/RejourneyNetworkEventFilter.swift
/**
 * Copyright 2026 Rejourney
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

enum RejourneyNetworkEventFilter {
    private static let lock = NSLock()
    private static let internalPathPrefixes = [
        "/api/sdk/config",
        "/api/ingest",
        "/upload/artifacts"
    ]
    private static let maxRegisteredInternalURLs = 200
    private static var apiBasePath = normalizeBasePath("https://api.rejourney.co")
    private static var registeredInternalURLs: [String] = []

    static func configure(apiURLString: String) {
        lock.lock()
        apiBasePath = normalizeBasePath(apiURLString)
        registeredInternalURLs.removeAll()
        lock.unlock()
    }

    static func registerInternalURL(urlString: String) {
        guard let normalized = normalizeURLString(urlString) else { return }
        lock.lock()
        if !registeredInternalURLs.contains(normalized) {
            registeredInternalURLs.append(normalized)
            if registeredInternalURLs.count > maxRegisteredInternalURLs {
                registeredInternalURLs.removeFirst(registeredInternalURLs.count - maxRegisteredInternalURLs)
            }
        }
        lock.unlock()
    }

    static func shouldIgnore(url: URL) -> Bool {
        if isRegisteredInternalURL(url) {
            return true
        }

        return shouldIgnore(path: url.path.isEmpty ? "/" : url.path)
    }

    static func shouldIgnore(details: [String: Any]) -> Bool {
        if let urlString = details["url"] as? String,
           let url = URL(string: urlString),
           shouldIgnore(url: url) {
            return true
        }

        if let path = details["urlPath"] as? String {
            return shouldIgnore(path: path)
        }

        return false
    }

    private static func shouldIgnore(path: String) -> Bool {
        let normalizedPath = normalizePath(path)
        let basePath = currentAPIBasePath()
        let prefixes: [String]
        if basePath.isEmpty {
            prefixes = internalPathPrefixes
        } else {
            prefixes = internalPathPrefixes + internalPathPrefixes.map { normalizePath("\(basePath)\($0)") }
        }

        return prefixes.contains { prefix in
            normalizedPath == prefix || normalizedPath.hasPrefix("\(prefix)/")
        }
    }

    private static func isRegisteredInternalURL(_ url: URL) -> Bool {
        guard let normalized = normalizeURL(url) else { return false }
        lock.lock()
        defer { lock.unlock() }
        return registeredInternalURLs.contains(normalized)
    }

    private static func currentAPIBasePath() -> String {
        lock.lock()
        defer { lock.unlock() }
        return apiBasePath
    }

    private static func normalizeBasePath(_ value: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "" }
        let url = URL(string: trimmed) ?? URL(string: "https://\(trimmed)")
        guard let path = url?.path, !path.isEmpty, path != "/" else { return "" }
        return normalizePath(path)
    }

    private static func normalizePath(_ value: String) -> String {
        let path = value.hasPrefix("/") ? value : "/\(value)"
        guard path.count > 1 else { return "/" }
        let trimmed = path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return trimmed.isEmpty ? "/" : "/\(trimmed)"
    }

    private static func normalizeURLString(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        guard let url = URL(string: trimmed) else {
            return trimmed.components(separatedBy: "#").first
        }
        return normalizeURL(url)
    }

    private static func normalizeURL(_ url: URL) -> String? {
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return url.absoluteString.components(separatedBy: "#").first
        }
        components.fragment = nil
        return components.string
    }
}
